File: Problem 6/C/MemoryEfficientDoublyLinkedList.c
#include <stdio.h>
#include <stdlib.h>

typedef struct
{
    int data;
    unsigned int both;
} Node;

Node *head = NULL, *tail = NULL;

const char *NULL_POINTER_ERROR = "\nUnable to allocate memory for Node : %d";
const char *REMOVE_ELEMENT_MSG = "\nRemoving element %d from doubly linked list.";
const char *EMPTY_LIST = "\nDoubly linked list is empty.";
const char *NEGATIVE_INDEX_ERROR = "\nNegative index is not possible.";

Node *createNode(int data)
{
    Node *node = (Node *)malloc(sizeof(Node));
    if (node != NULL)
    {
        node->data = data;
        node->both = 0;
    }
    return node;
}

int insertFirst(int data)
{
    Node *temp = createNode(data);
    if (temp == NULL)
    {
        printf(NULL_POINTER_ERROR, data);
        return 0;
    }

    if (head == NULL)
    {
        head = temp;
        tail = temp;
        return 1;
    }
    else
    {
        unsigned int nextAdd = (unsigned int)head;
        unsigned int prevAdd = (unsigned int)temp;
        temp->both ^= nextAdd;
        head->both ^= prevAdd;
        head = temp;
        return 1;
    }

    printf("\nNot able to insert Node : %d at beginning to doubly linked list", data);

    free(temp);
    return 0;
}

int insertLast(int data)
{
    Node *temp = createNode(data);
    if (temp == NULL)
    {
        printf(NULL_POINTER_ERROR, data);
        return 0;
    }
    if (head == NULL)
    {
        head = temp;
        tail = temp;
        return 1;
    }
    else
    {
        unsigned int nextAdd = (unsigned int)temp;
        unsigned int prevAdd = (unsigned int)tail;
        tail->both ^= nextAdd;
        temp->both ^= prevAdd;
        tail = temp;
        return 1;
    }

    printf("\nNot able to insert Node : %d at last to doubly linked list", data);
    free(temp);
    return 0;
}

int insertAt(int index, int data)
{
    int count = 0;
    if (head == NULL)
    {
        printf(EMPTY_LIST);
        return 0;
    }
    if (index < 0)
    {
        printf(NEGATIVE_INDEX_ERROR);
        return 0;
    }
    if (index == 0)
    {
        return insertFirst(data);
    }
    Node *temp = createNode(data);

    if (temp == NULL)
    {
        printf(NULL_POINTER_ERROR, data);
        return 0;
    }
    Node *ptr = head;
    unsigned int prevAdd = 0;
    unsigned int nextAdd = 0;
    while (ptr != NULL && count != index)
    {
        nextAdd = prevAdd ^ ptr->both;
        prevAdd = (unsigned int)ptr;
        ptr = (Node *)nextAdd;
        count++;
    }
    if (index == count)
    {
        if (ptr != NULL)
        {
            ((Node *)prevAdd)->both ^= (unsigned int)ptr ^ (unsigned int)temp;
            temp->both = prevAdd ^ (unsigned int)ptr;
            ptr->both ^= prevAdd ^ (unsigned int)temp;
            return 1;
        }
        else
        {
            return insertLast(data);
        }
    }

    printf("\nNot able to insert Node : %d at index : %d", data, index);
    free(temp);
    return 0;
}

void traversDoublyLinkedList()
{
    Node *ptr = head;
    unsigned int prevAdd = 0;
    if (head == NULL)
    {
        tail = head;
        printf(EMPTY_LIST);
        return;
    }
    printf("\nElement in doubly linked list :");
    while (ptr != NULL)
    {
        printf(" %d ", ptr->data);
        unsigned int temp = (unsigned int)ptr;
        ptr = (Node *)(ptr->both ^ prevAdd);
        prevAdd = temp;
    }
}

void traversDoublyLinkedListReverse()
{
    Node *ptr = tail;
    unsigned int nextAdd = 0;
    if (tail == NULL)
    {
        head = tail;
        printf(EMPTY_LIST);
        return;
    }
    printf("\nElement in doubly linked list (reverse order) :");
    while (ptr != NULL)
    {
        printf(" %d ", ptr->data);
        unsigned int temp = (unsigned int)ptr;
        ptr = (Node *)(ptr->both ^ nextAdd);
        nextAdd = temp;
    }
}

int size()
{
    Node *ptr = head;
    int count = 0;
    unsigned int prevAdd = 0;
    while (ptr != NULL)
    {
        count++;
        unsigned int temp = (unsigned int)ptr;
        ptr = (Node *)(ptr->both ^ prevAdd);
        prevAdd = temp;
    }
    return count;
}

int removeFirst()
{
    if (head == NULL)
    {
        printf(EMPTY_LIST);
        return 0;
    }

    printf(REMOVE_ELEMENT_MSG, head->data);

    if (head == tail)
    {
        free(head);

        head = NULL;
        tail = NULL;
        return 1;
    }
    else
    {
        Node *temp = head;
        head = (Node *)temp->both;
        head->both ^= (unsigned int)temp;

        free(temp);
        return 1;
    }
    printf("\nNot able to remove 1st element from doubly liked list.");
    return 0;
}

int removeLast()
{
    if (head == NULL)
    {
        printf(EMPTY_LIST);
        return 0;
    }

    printf(REMOVE_ELEMENT_MSG, tail->data);

    if (head == tail)
    {
        free(head);

        head = NULL;
        tail = NULL;
        return 1;
    }
    else
    {
        Node *temp = tail;
        tail = (Node *)temp->both;
        tail->both ^= (unsigned int)temp;

        free(temp);
        return 1;
    }

    return 0;
}

int removeAt(int index)
{
    if (index < 0)
    {
        printf(NEGATIVE_INDEX_ERROR);
        return 0;
    }

    if (head == NULL)
    {
        printf(EMPTY_LIST);
        return 0;
    }
    if (index == 0)
    {
        return removeFirst();
    }
    int count = 0;
    Node *ptr = head;
    unsigned int prevAdd = 0;
    unsigned int nextAdd = 0;
    while (ptr != NULL && count != index)
    {
        nextAdd = prevAdd ^ ptr->both;
        prevAdd = (unsigned int)ptr;
        ptr = (Node *)nextAdd;
        count++;
    }
    if (index == count)
    {
        if (ptr != NULL)
        {
            printf(REMOVE_ELEMENT_MSG, ptr->data);
            nextAdd = ptr->both ^ prevAdd;
            ((Node *)prevAdd)->both ^= (unsigned int)ptr ^ nextAdd;
            ((Node *)nextAdd)->both ^= (unsigned int)ptr ^ prevAdd;
            free(ptr);
            return 1;
        }
        else
        {
            return removeLast();
        }
    }

    printf("\nNot able to remove Node at index : %d", index);
    return 0;
}

void clear()
{
    Node *ptr = head;
    unsigned int prevAdd = 0;
    if (head == NULL)
    {
        tail = head;
        printf(EMPTY_LIST);
        return;
    }
    while (ptr != NULL)
    {
        unsigned int temp = (unsigned int)ptr;
        ptr = (Node *)(ptr->both ^ prevAdd);
        prevAdd = temp;
        free((Node *)temp);
    }
    head = tail = NULL;
    printf("\n Deleted doubly linked list.");
}

int *get(int index)
{
    if (index < 0)
    {
        printf(NEGATIVE_INDEX_ERROR);
        return NULL;
    }
    Node *ptr = head;
    int count = 0, *temp = NULL;
    unsigned int prevAdd = 0;
    while (ptr != NULL && count != index)
    {
        count++;
        unsigned int temp = (unsigned int)ptr;
        ptr = (Node *)(ptr->both ^ prevAdd);
        prevAdd = temp;
    }
    if (count == index)
    {
        if (ptr != NULL)
        {
            temp = (int)malloc(sizeof(int));
            *temp = ptr->data;
        }
    }
    if (temp == NULL)
        printf("\nIndex %d not possible in doubly linked list.", index);
    return temp;
}

int main()
{
    int ch, data, index, *x, s;
    const char *ENTER_ELEMENT = "\n Enter element to add at %s : ";
    const char *SEPERATOR = "\n======================================================================\n";
    do
    {
        printf(SEPERATOR);
        printf("\n* * * * * * * MENU * * * * * * *\n");
        printf("\n 1.  Insert First\n 2.  Insert Last\n 3.  Insert At Index");
        printf("\n 4.  Remove First\n 5.  Remove Last\n 6.  Remove From Index");
        printf("\n 7.  Get Element At Index\n 8.  Size\n 9.  Clear \n 10. Display\n 11. Display Reverse");
        printf("\n 12. Exit\nEnter Your Choice : ");

        scanf("%d", &ch);
        switch (ch)
        {
        case 1:
            printf(SEPERATOR);
            printf(ENTER_ELEMENT, "beginning");
            scanf("%d", &data);
            insertFirst(data);
            break;

        case 2:
            printf(SEPERATOR);
            printf(ENTER_ELEMENT, "last");
            scanf("%d", &data);
            insertLast(data);
            break;

        case 3:
            printf(SEPERATOR);
            printf("\n Enter index : ");
            scanf("%d", &index);
            char str[16]; //= "index-";
            sprintf(str,"index-%d", index);
            printf(ENTER_ELEMENT, str);
            scanf("%d", &data);
            insertAt(index, data);
            break;

        case 4:
            printf(SEPERATOR);
            removeFirst();
            break;

        case 5:
            printf(SEPERATOR);
            removeLast();
            break;

        case 6:
            printf(SEPERATOR);
            printf("\n Enter index : ");
            scanf("%d", &index);
            removeAt(index);
            break;

        case 7:
            printf(SEPERATOR);
            printf("\n Enter index : ");
            scanf("%d", &index);
            x = get(index);
            if (x != NULL)
            {
                printf("\n Element at index : %d is %d.", index, *x);
            }
            break;

        case 8:
            printf(SEPERATOR);
            s = size();
            printf("\n Size of doubly linked list is : %d", s);
            break;

        case 9:
            printf(SEPERATOR);
            clear();
            break;

        case 10:
            printf(SEPERATOR);
            traversDoublyLinkedList();
            break;

        case 11:
            printf(SEPERATOR);
            traversDoublyLinkedListReverse(index);
            break;

        case 12:
            break;

        default:
            printf("\n INVALID CHOICE......!");
            break;
        }
    } while (ch != 12);

    return 0;
}

File: Problem 6/JAVA/MemoryEfficientDoublyLinkedList.java
import java.util.HashMap;
import java.util.Map;

/**
 * @author Ajay G.
 *
 */
class Node {
	private final static Map<Integer, Node> NODES_DATA_MAP = new HashMap<>();

	private int data;
	private int both;

	/**
	 * 
	 */
	public Node(int data) {
		this.data = data;
		Node.NODES_DATA_MAP.put(this.hashCode(), this);
		this.both = 0;
	}

	public int getBoth() {
		return both;
	}

	public void setBoth(int both) {
		this.both = both;
	}

	public int getData() {
		return data;
	}

	public void setData(int data) {
		this.data = data;
	}

	public static Map<Integer, Node> getNodesDataMap() {
		return NODES_DATA_MAP;
	}

	public void clear() {
		Node.NODES_DATA_MAP.clear();
	}

}

public class MemoryEfficientDoublyLinkedList {

	private final static String REMOVE_ELEMENT_MSG = "\nRemoving element %d from doubly linked list.";
	private final static String EMPTY_LIST = "\nDoubly linked list is empty.";
	private final static String NEGATIVE_INDEX_ERROR = "\nNegative index is not possible.";

	private Node head, tail;

	/**
	 * 
	 */
	public MemoryEfficientDoublyLinkedList() {
		this.tail = null;
		this.head = null;
	}

	int insertFirst(int data) {
		Node temp = new Node(data);
		if (this.head == null) {
			this.head = temp;
			this.tail = temp;
			return 1;
		} else {
			int nextAdd = this.head.hashCode();
			int prevAdd = temp.hashCode();
			temp.setBoth(temp.getBoth() ^ nextAdd);
			this.head.setBoth(this.head.getBoth() ^ prevAdd);
			this.head = temp;
			return 1;
		}
	}

	int insertLast(int data) {
		Node temp = new Node(data);
		if (this.head == null) {
			this.head = temp;
			this.tail = temp;
			return 1;
		} else {
			int nextAdd = temp.hashCode();
			int prevAdd = this.tail.hashCode();
			this.tail.setBoth(this.tail.getBoth() ^ nextAdd);
			temp.setBoth(temp.getBoth() ^ prevAdd);
			this.tail = temp;
			return 1;
		}
	}

	int insertAt(int index, int data) {
		int count = 0;
		if (this.head == null) {
			System.out.print(EMPTY_LIST);
			return 0;
		}
		if (index < 0) {
			System.out.print(NEGATIVE_INDEX_ERROR);
			return 0;
		}
		if (index == 0) {
			return insertFirst(data);
		}
		Node temp = new Node(data);

		Integer ptr = this.head.hashCode();
		Integer prevAdd = 0;
		Integer nextAdd = 0;
		while (ptr != null && count != index) {
			Node ptrNode = Node.getNodesDataMap().get(ptr);
			nextAdd = prevAdd ^ ptrNode.getBoth();
			prevAdd = ptr;
			ptr = nextAdd;
			count++;
		}
		if (index == count) {
			if (ptr != null) {
				Node prevNode = Node.getNodesDataMap().get(prevAdd);
				prevNode.setBoth(prevNode.getBoth() ^ ptr ^ temp.hashCode());
				temp.setBoth(prevAdd ^ ptr);
				Node ptrNode = Node.getNodesDataMap().get(ptr);
				ptrNode.setBoth(ptrNode.getBoth() ^ prevAdd ^ temp.hashCode());
				return 1;
			} else {
				return insertLast(data);
			}
		}

		System.out.print("\nNot able to insert Node : " + data + " at index : " + index);
		return 0;
	}

	void traversDoublyLinkedList() {
		Integer ptr = this.head.hashCode();
		int prevAdd = 0;
		if (this.head == null) {
			this.tail = this.head;
			System.out.print(EMPTY_LIST);
			return;
		}
		System.out.print("\nElement in doubly linked list :");
		Node ptrNode = Node.getNodesDataMap().get(ptr);
		while (ptrNode != null) {

			System.out.print(" " + ptrNode.getData() + " ");
			int temp = ptr;
			ptr = ptrNode.getBoth() ^ prevAdd;
			prevAdd = temp;
			ptrNode = Node.getNodesDataMap().get(ptr);
		}
	}

	void traversDoublyLinkedListReverse() {
		Integer ptr = this.tail.hashCode();
		int nextAdd = 0;
		if (this.tail == null) {
			this.head = this.tail;
			System.out.print(EMPTY_LIST);
			return;
		}
		System.out.print("\nElement in doubly linked list (reverse order) :");
		Node ptrNode = Node.getNodesDataMap().get(ptr);
		while (ptrNode != null) {
			System.out.print(" " + ptrNode.getData() + " ");
			int temp = ptr;
			ptr = ptrNode.getBoth() ^ nextAdd;
			nextAdd = temp;
			ptrNode = Node.getNodesDataMap().get(ptr);
		}
	}

	int size() {
		return Node.getNodesDataMap().size();
	}

	int removeFirst() {
		if (this.head == null) {
			System.out.print(EMPTY_LIST);
			return 0;
		}

		System.out.print(String.format(REMOVE_ELEMENT_MSG, this.head.getData()));

		if (this.head == this.tail) {
			Node.getNodesDataMap().remove(this.head.hashCode());
			this.head = null;
			this.tail = null;
			return 1;
		} else {
			Integer temp = this.head.hashCode();
			Node temNode = Node.getNodesDataMap().get(temp);
			this.head = Node.getNodesDataMap().get(temNode.getBoth());
			this.head.setBoth(this.head.getBoth() ^ temp);

			Node.getNodesDataMap().remove(temp);
			return 1;
		}
	}

	int removeLast() {
		if (this.head == null) {
			System.out.print(EMPTY_LIST);
			return 0;
		}

		System.out.print(String.format(REMOVE_ELEMENT_MSG, this.tail.getData()));

		if (this.head == this.tail) {
			Node.getNodesDataMap().remove(this.head.hashCode());

			this.head = null;
			this.tail = null;
			return 1;
		} else {
			Integer temp = this.tail.hashCode();
			Node tempNode = Node.getNodesDataMap().get(temp);
			this.tail = Node.getNodesDataMap().get(tempNode.getBoth());
			this.tail.setBoth(this.tail.getBoth() ^ temp);

			Node.getNodesDataMap().remove(temp);
			return 1;
		}
	}

	int removeAt(int index) {
		if (index < 0) {
			System.out.print(NEGATIVE_INDEX_ERROR);
			return 0;
		}

		if (this.head == null) {
			System.out.print(EMPTY_LIST);
			return 0;
		}
		if (index == 0) {
			return removeFirst();
		}
		int count = 0;
		Integer ptr = this.head.hashCode();
		Node ptrNode = Node.getNodesDataMap().get(ptr);
		int prevAdd = 0;
		int nextAdd = 0;
		while (ptrNode != null && count != index) {
			nextAdd = prevAdd ^ ptrNode.getBoth();
			prevAdd = ptr;
			ptr = nextAdd;
			count++;
			ptrNode = Node.getNodesDataMap().get(ptr);
		}
		if (index == count) {
			if (ptrNode != null) {
				System.out.print(String.format(REMOVE_ELEMENT_MSG, ptrNode.getData()));
				Node prevNode = Node.getNodesDataMap().get(prevAdd);
				nextAdd = ptrNode.getBoth() ^ prevAdd;
				prevNode.setBoth(prevNode.getBoth() ^ ptr ^ nextAdd);
				Node nextNode = Node.getNodesDataMap().get(nextAdd);
				nextNode.setBoth(nextNode.getBoth() ^ ptr ^ prevAdd);
				Node.getNodesDataMap().remove(ptr);
				return 1;
			} else {
				return removeLast();
			}
		}

		System.out.print("\nNot able to remove Node at index : " + index);
		return 0;
	}

	void clear() {
		Node.getNodesDataMap().clear();
		this.head = this.tail = null;
		System.out.print("\n Deleted doubly linked list.");
	}

	Integer get(int index) {
		if (index < 0) {
			System.out.print(NEGATIVE_INDEX_ERROR);
			return null;
		}
		Integer ptr = this.head.hashCode();
		Node ptrNode = Node.getNodesDataMap().get(ptr);
		int count = 0;
		Integer temp = null;
		Integer prevAdd = 0;
		while (ptrNode != null && count != index) {
			count++;
			temp = ptr;
			ptr = ptrNode.getBoth() ^ prevAdd;
			prevAdd = temp;
			ptrNode = Node.getNodesDataMap().get(ptr);
		}
		if (count == index) {
			if (ptr != null) {
				temp = ptrNode.getData();
			}
		}
		if (temp == null)
			System.out.print("\nIndex " + index + " not possible in doubly linked list.");
		return temp;
	}
}
